exercises/react-context/flow.txt

```diff
@@ -0,0 +1,147 @@
+the username and password in a JSON string.
+-The server will get the request and pass it to a function that has the matching
+http method and request url.
+-First, the function will check if there is a username and password in the req.body.
+If there isn't a username and password in the body, the function will throw a new
+ClientError which will send a status code of 401 and the message 'invalid login'.
+This error should never trigger since the form requires those fields to be filled
+out before submitting.
+-Assuming that there is a username and password in the req.body, the database will
+be queried. The sql will select the userId and hashedPassword from the users table
+and get the ones where the username is equal to the provided username from the user.
+-If the database doesn't return a response meaning that there is no user with that
+username, it will throw a new ClientError with the status code of 401 and the
+message 'invalid login'.
+-If there is a response, the given password from the user and the already hashed
+password from the data base is compared by the verify method of the argon2 object.
+-Then the respnse from the verify method is checked to see if it came truthy or
+falsey.
+-If the passwords did not match, a new new ClientError with the status code of 401
+and the message 'invalid login' will be thrown.
+-If the passwords do match, the userId and password will be assigned as properties
+and values to the const variable payload.
+-The sign method of the jwt object will be called passing in the payload variable
+and the process.evn.TOKEN_SECRET and assigned to the const variable token.
+-An object containing the properties token and user and the values of token and
+payload accordingly are sent as the response in JSON from the server.
+-The then method after the fetch method will get the response and then parse the
+response.
+-Once the response is parsed, the hash location will be checked and since the user
+was signing in, and there is a user and token property of the user object, the
+onSignIn method will be called with the result as the argument.
+-The handleSignIn method will be triggered and the setItem method of the localStorage
+property of the window object will be called with the arguments 'react-context-jwt'
+and token. This will store the token in the local storage on the user's device.
+-The state will also be set with the object containing the user property and the
+value of the user variable that was destructured from the result.
+-Since the state was changed, the page will be rendered again, but this time the
+state will contain the value the user.
+-The renderPage method will be called and will check the route property of the
+state property of the this object. Since the route is still sign-in, The Auth
+react element will be rendered. But after the component mounts, the getItem method
+of the localStorage property of the window object is called with the argument
+'react-context-jwt' and assigned to token.
```

Comment:
Close. `componentDidMount` only runs one time though, after the first render. So during this phase, the Auth component is rendered again, but it returns a Redirect element instead.